### PR TITLE
[READY] Adds ability to give airlocks one-way unrestricted access

### PR DESCRIPTION
**Pull request by zxaber**
:cl: Zxaber
add: Airlock electronics can have now have unrestricted access by direction set. The resulting airlock will allow all traffic from the specified direction(s) while still requiring normal access otherwise. A small floor light will indicate this.
tweak: Medbay Cloning and Main Access doors now have unrestricted access settings set, and the buttons have been removed. All maps have been updated.
fix: Airlocks now correctly update their overlays (bolts lights, emergency lights) when their power state changes.
/:cl:

The Medbay door access, with the buttons that allow free exit, always felt a bit clunky. And in general, it seemed strange to me that a door that you don't have access to can keep you from leaving areas and getting back to the main hallway (though it makes sense in high-security areas). So I made a system that allows doors to give unrestricted access based on the location of the user.

A video, for better explanation; https://youtu.be/mNavaV-QKJM.

Diagonal access is intentionally inherited from neighbor cardinals, because it allows people to open doors they're not directly in front of (like people naturally do already), and because I can't think of a single time it'll cause problems.

This PR only adds the ability to construct new doors with this feature. Assuming it gets merged, I'll put out a following PR that updates the maps.

**Update 7-22:** Right, so here's where it's at. Cyprex is right in that I should do the full-map changes in their on PRs to allow for the map balance discussion (which was my original intention anyway). MSO requested an indicator system before this gets merged in a way that players can build with it, so that people don't build traps in maint. And I didn't want to only make changes to the medbay doors without any indicator anyway to avoid the confusion about why the button is gone and why these doors arbitrarily allow access from one side. With that in mind, I added some basic indicators in the form of a small white light that shines in the available directions. It's small enough that even if a wall is placed up against it, it doesn't look too terrible. Since it's supposed to be a floor-facing light, airlocks now emit a tiny bit of light when any side is set to unrestricted access.

While I was at it, I fixed door overlays not getting updated with power state changes properly. There was already code in place to check the power state, so things like bolt lights turning off with the power were intended at some point, but no trigger was in place unless someone popped open a panel or something. The overlays now respond to power changes correctly, be it due to area power loss or a wire being cut. If a door has a set light level due to the unres floor light, it is also turned off while the power is out.

I've tested these changes, but it's possible I've missed something while copying from my local files to github, so I will test a fresh download of the branch soon.

unres_indicator

**Edit 7-31:** Everything is tested and working. Unless further changes are requested, I think this PR is ready.

### CODE REVIEW

**Comment by epicmemename** [comment]
speedmerge

**Review by ChangelingRain — DISMISSED** [review]
nice

File: code/game/machinery/doors/door.dm [hunk 1]
```diff
@@ -163,8 +164,23 @@
 /obj/machinery/door/allowed(mob/M)
 	if(emergency)
 		return TRUE
+	if(unrestricted_side(M))
+		return TRUE
 	return ..()
 
+/obj/machinery/door/proc/unrestricted_side(mob/M) //Allows for specific side of airlocks to be unrestrected (IE, can exit maint freely, but need access to enter)
+	while(!istype(M.loc, /turf)) //If user is in an object (like a mech), we need to use the object's X and Y.
```
Comment (SpaceManiac) [inline code/game/machinery/doors/door.dm]: just use `get_turf`

File: code/game/machinery/doors/door.dm [hunk 2]
```diff
@@ -163,8 +164,23 @@
...
+		M = M.loc
+	if(unres_sides & 1 && M.y == y + 1) //1 is North
+		return TRUE
+	if(unres_sides & 2 && M.x == x + 1) //2 is East
+		return TRUE
+	if(unres_sides & 4 && M.y == y - 1) //4 is South
+		return TRUE
+	if(unres_sides & 8 && M.x == x - 1) //8 is West
+		return TRUE
```
Comment (SpaceManiac) [inline code/game/machinery/doors/door.dm]: Actually you could replace this whole function with `get_dir(src, M) & unres_sides`

File: code/game/machinery/doors/airlock_electronics.dm [hunk 3]
```diff
@@ -55,3 +57,7 @@
 			else
 				accesses -= access
 			. = TRUE
+		if("direc_set")
+			var/unres_direction = text2num(params["unres_direction"])
+			unres_sides = unres_sides ^ unres_direction //XOR, toggles only the bit that was clicked
```
Comment (SpaceManiac) [inline code/game/machinery/doors/airlock_electronics.dm]: `unres_sides ^= unres_direction`

**Review by SpaceManiac — CHANGES_REQUESTED** [review]
Nitpicks, useful feature for engineers to play with even if it ends up not being applied to the medbay doors for w/e reason

**Comment by Cyberboss** [comment]
Top tier pr

**Comment by MrStonedOne** [comment]
I'd like see to see some design notes in the wiki or any mapping guides, strongly recommending this only be applied in the direction going from a restricted to a common area, and never the reverse.

Common sense, until its not.

**Comment by MrStonedOne** [comment]
Also, we might need some sort of indicator that this is the case, otherwise its an easy pickings trap to apply this to any of the fire fighter maint rooms to trap people in places with no easy way to get out since no ai view and its hard to explain where you are when you're in a random 3x2 room in maint.

**Comment by Jalleo** [comment]
^ That now makes me think of [Insert TV drama where everyone ends up getting locked in a room for a episode one person at a time]

**Comment by zxaber** [comment]
Y'know, I had actually thought of a clown prank using this involving four airlocks in a diamond shape with the center empty, so that one could enter from the sides but (if they stuck around long enough) not leave. Still, it's a lot less fun to be trapped and forgotten in maintenance. 

I haven't touched adding sprites yet, so I'll have to look into that. As for actual indicator design, my first thought was lighting up a side of the airlock's face. But I'm not sure I like how it feels like an indicator to the player rather than a conceivable in-universe display  . I suppose it's sorta hologram-y. But we already have bolt lights, emergency lights, and electronics panel on the face. Aside from that, I could try placing an overlay onto the bordering tile, as if a floor light was shining down. A wide one would be more noticeable, but a narrow one might be cleaner looking.

Anyway, quick and dirty mock-up of possible indicators.
image
I'm certainly open to suggestions for better ideas.

**Comment by nicbn** [comment]
Add a yellow blinking light on the floor, rounded around the center. Example:

image

This one is a FF0 100% to FF0 0% radial gradient going from the right-most of the airlock to the center of the floor tile.

**Comment by Jalleo** [comment]
I feel that wouldn't work correctly when the room then goes dark due to lighting going out.

**Comment by zxaber** [comment]
I want to avoid linking this feature with emergency access, since emergency access can be disabled at any moment, and already has it's own indicator anyway. That said, I figured it was also a good idea to avoid yellow (and red) for similar reasons. I had used blue, because it wasn't used in door lights yet, but green could work since it links to the "access granted" idea.

**Comment by Shdorsh** [comment]
This is great and would do away with buttons in cloning and medbay. Just a few questions:

Do diagonals still get their emergency access inherited from a cardinal if that cardinal is facing a wall?

**Comment by zxaber** [comment]
Currently yes, as I didn't see a reason to specially check for that. I suppose it would let you build "secret entrance" rooms where you can enter without an ID if you know where to stand. Depends how the indicators end up working.

**Comment by Ratvar** [comment]
Maybe make it so you can define access for both ends?

**Comment by zxaber** [comment]
I feel like setting multiple access levels is getting above and beyond the purpose of this PR, and indeed would be much more complex than it would be useful (both in code and in use). The point of this PR is to let people without access get back to the open hallways. For the freshly-cloned crew member with no ID; for the assistant that wandered into maintenance and got stuck; for the janitor that was let in to clean Robotics and now is waiting to be let back out. If we're looking at letting each side have arbitrary but defined levels of access, we should probably first ask why a single door is being used for different access levels, and why crew that are expected to be in both sides at some point can't just be given access to both.

In other news, I ran into a snag with overlays, in that I can't figure out a clean way to remove them if something like a wall gets built over it. I wanted to go with the small arrow-shaped floor light, but it's gonna be terrible if that just sits on top of a wall. I tried to see if using decals was at all viable, but it's really a sort of hacky solution on it's own, and removing a wall that was placed over a decal removes the decal anyway, which is basically the same issue as before. So I'll try overlaying a tile-wide and thin glow, which shouldn't be horrible against a wall (I think). The other issue is that my overlays aren't showing up immediately, and I think it's because the first overlay cycle is done before my unres_sides variable gets moved to the door. Once I figure that fix out, I'll update the PR.

**Comment by nicbn** [comment]
If you're going to add the in-floor indicators in this PR add the WIP tag to it

**Comment by MrStonedOne** [comment]
The best solution is to just make this mapper only for now then add the overlays and in game way of making it later.

**Comment by zxaber** [comment]
So, switching gears here and moving to the map changes. Since player-built doors won't have this yet, I might as well do the map changes. I've added unrestricted access settings to doors across Metastation under the following guidelines;

- Unrestricted access is used to allow crew to get back the common areas.
- Airlocks should not allow open access _into_ maintenance, unless no other option exists. 
- Likewise, no open access _into_ departments from maintenance.
- Exit from a department should in general use the department's main entrance, unless the walk to a side entrance involves fewer doors.
- High security airlocks do not have unrestricted access. This includes all security areas and checkpoints, all heads of staff (and QM) offices, the Bridge as a whole, potentially-dangerous areas (tech storage, toxins storage) and areas that are considered critical to the station health (Atmos, the Supermatter chamber, any room that has an SMES). Access doors leading to and on the AI's satellite also do not have unrestricted access settings.
- Containment areas do not have unrestricted access. This includes Xeno access, Viro access, and the mining wing.
- Outer airlocks do not have unrestricted access. (the exception is the auxiliary building airlock and the engineering escape pod airlock).


A diagram of the changes. Ignore white arrows.
metastation-unres

Comments welcome, etc. If there are no objections to this, I will apply similar changes to Delta, Box, Pubby, and Omega. Also, this is being done only to the station proper. The mining outpost falls under a containment area (in my opinion). The buttons will be commented out in the airlock electronics tgui file, until I get an indicator system working in a later PR.

**Comment by Cyprex** [comment]
Good, but I'd PR the mapping changes seperately, as they could have significant impact on gameplay(Harder to trap people and such) and are open for discussion. Finally, I believe the medbay surgery spectator room is public already?

**Comment by AzureWildfire** [comment]
I like the idea, I have some recommendations about the map changes if you end up doing them: 
Leave doors like _[Engineering (except engineering common area), the maint outside captains office, Main cargo (not disposals and entrance those are fine (I think)), genetics (maybe), toxins (is better to leave the maint as an exit that let someone into toxins),  maint surgery, maybe other places]_ in normal mode, those are doors that you need to break in to be in there and if they broke in they shouldn't have an easy way out (I know that they also can be inside because of a bluespace anomaly).
**Basically don't give exit access in places that people shouldn't be in.**

**Comment by epochayur** [comment]
this is one of those PRs that I didn't know we needed so badly until it got pulled

**Comment by JJRcop** [comment]
You should install a git client so you can upload your changes directly instead of having to repeat them with the GitHub web editor. Check out https://help.github.com/articles/set-up-git/ to start, and visit [#coderbus on rizon.net](https://qchat.rizon.net/?channels=coderbus) for any further questions.

**Comment by zxaber** [comment]
@AzureWildfire
Yeah, I might end up rethinking the station-wide changes when I do those PRs. Most places do have cameras, so the AI can help out. I'll probably just make it a maintenance thing, since free access to maint can be enabled and disabled by command staff at will, and there's rarely any cameras to help the AI let people get out.

**Comment by Cyprex** [comment]
Consider submitting the mapping changes in a separate PR after this is done.

**Comment by zxaber** [comment]
@Cyprex
I'd like to include the medical doors here, but the map-wide changes will be their own PR.

**Comment by Enzarel** [comment]
While i love the idea of fixing medbay doors, i have objections to some of your proposed placements, for example:
Engineering doors allow for hit-and-run greytiding
Chef can no longer safely murder people in his backroom. Also pete can get out.
Surgery has the same problem.
Breaking into south medbay now requires hacking only one airlock
You see that small aux botany over xeno? Now look at the airlock. Now imagine a cult teleport in that room. It's already overused as a cult place.
Getting into toxins from the back requires hacking one door.
More hit-and-run greytiding all over science.
I see no logical way a person can get into the science server room. If they get there with some """help""", they should most likely stay there. We should encourage creative tators.
Cargo is already a greyfest, so no objections here.

**Review by ShizCalev — CHANGES_REQUESTED** [review]
have `unres_sides` be set by a airlock mapping helper instead of directly via the airlock's vars (see `/obj/effect/mapping_helpers/airlock/cyclelink_helper` for an example), that way we can quickly see changes related to them in map diffs, and mappers can easily see incorrect usages without having to manually check every instance.

**Comment by zxaber** [comment]
Added helpers, they look like this;
image

Will update the maps to use helpers rather than var edits soon.

**Comment by zxaber** [comment]
Either Dreammaker or Map Merge is being weird, and reshuffling the entire map files when I try to make any changes, leading to a lot of add/remove text that shouldn't be there. I can just notepad-edit the files, and I'll take care of that later today. Removing the ready tag for the moment.

File: code/game/machinery/doors/airlock.dm [hunk 4]
```diff
@@ -584,6 +589,31 @@
 	if((!(. = airlock_overlays[iconkey])))
 		. = airlock_overlays[iconkey] = mutable_appearance(icon_file, icon_state)
 
+/obj/machinery/door/airlock/proc/check_unres() //unrestricted sides. This overlay indicates which directions the player can access even without an ID
+	if(hasPower() && unres_sides)
+		if(src.unres_sides & 1)
```
Comment (AnturK) [inline code/game/machinery/doors/airlock.dm]: Please use direction defines in these and src is unnecessary.

**Review by AnturK — CHANGES_REQUESTED** [review]
Just some minor nitpick while i testmerge this.

File: code/game/machinery/doors/airlock.dm (hunk 4 above)
Comment (zxaber) [reply code/game/machinery/doors/airlock.dm]: By directional defines, do you mean setting specifically to North, East, etc rather than using bitflags?

**Comment by nicbn** [comment]
You have to prepare maps after you map merge a change so it won't reshuffle it in the next time you run map merge

File: code/game/machinery/doors/airlock.dm (hunk 4 above)
Comment (AnturK) [reply code/game/machinery/doors/airlock.dm]: Just mean `unres_sides & NORTH` instead of `unres_sides & 1`

File: code/game/machinery/doors/airlock.dm [hunk 5]
```diff
@@ -41,6 +41,11 @@
 #define ELECTRIFIED_PERMANENT -1
 #define AI_ELECTRIFY_DOOR_TIME 30
 
+#define NORTH 1 //defines for the unrestricted sides (unres) section
```
Comment (AnturK) [inline code/game/machinery/doors/airlock.dm]: No need for these, you're just duplicating the ones in stddef.dm

**Review by AnturK — CHANGES_REQUESTED** [review]

**Comment by zxaber** [comment]
Issue found with the RPD and tgui changes. Narrowing down, will try to fix later today.

**Comment by AnturK** [comment]
tgui build up to date now ?

**Review by ShizCalev — CHANGES_REQUESTED** [review]

**Comment by Iamgoofball** [comment]
You need to update and recompile tgui to fix the conflict.

**Comment by zxaber** [comment]
I did that to fix the RPD issue. Are there other issues showing up in the test merge?

**Comment by AnturK** [comment]
Another merged PRupdated tgui, just needs update. There's no other issues afaik.

**Comment by zxaber** [comment]
I see. I'm out of town on vacation this week, but I'll see if I can remote into my computer at some point and take care of it.

**Review by SpaceManiac — DISMISSED** [review]

**Comment by epochayur** [comment]
WOOOOOO

**Comment by zxaber** [comment]
Hey, thanks a ton for fixing the conflict and getting this merged.

**Comment by nicbn** [comment]
they actually merge it the absolute madmen
